ddtrace/contrib/trace_utils.py

```diff
@@ -102,13 +120,28 @@ def _store_headers(headers, span, integration_config, request_or_response):
         return
 
     for header_name, header_value in headers.items():
+        """config._header_tag_name gets an element of the dictionary in config.http._header_tags
+        which gets the value from DD_TRACE_HEADER_TAGS environment variable."""
         tag_name = integration_config._header_tag_name(header_name)
         if tag_name is None:
             continue
         # An empty tag defaults to a http.<request or response>.headers.<header name> tag
         span.set_tag(tag_name or _normalize_tag_name(request_or_response, header_name), header_value)
 
 
+def _get_request_header_user_agent(headers):
+    # type: (Dict[str, str], Span, IntegrationConfig, str) -> str
+    """Get user agent from request headers
+    :param headers: A dict of http headers to be stored in the span
+    :type headers: dict or list
+    """
+    for key_pattern in USER_AGENT_PATTERNS:
```

Comment:
🤔 I am worried about the performance impact of this. it should be worse case `O(n)` where today `n = 18`, but 18 dict gets _could_ (hypothesis) be a little heavy on every web request.

It might be worth us adding performance benchmarks for `set_http_meta` first so we can compare before/after impact.

What do you think?

https://github.com/DataDog/dd-trace-py/tree/1.x/benchmarks


We also have some basic benchmarks for Django and Flask, but we've found them to have mixed results (no performance difference whether tracing is enabled or not, which is not accurate).

Reply:
I know we are being this exhaustive to be sure we don't miss it, but aren't there only a finite set of possible values?

We should probably have either WSGI (`HTTP_USER_AGENT`) or do most frameworks offer case insensitive lookups? (I know Flask/Werkzeug is case insensitive, Django has `request.META` which has `HTTP_USER_AGENT`, etc).

I wonder if we need this to be this exhaustive, or if we can/should optimize for known best case scenarios first?

Reply:
At first look could be weird, but I measured the performance of this PR (and other failed attempts) and a list checking the key is faster than regex:
https://gist.github.com/avara1986/07a312d7ca5e2a6e1547eb60c2cd6ca3

Common cases, this list is x10 faster than regex

I like your idea about benchmarks :smiley: what do you think about below plan: 
- WIP This PR 
- Open new PR with benchmark tests on set_http_meta 
- Measure this PR
I like your idea about benchmarks :smiley: what do you think about below plan: 
- WIP This PR 
- Open new PR with benchmark tests on set_http_meta 
- Measure this PR

Reply:
that plan sounds good. a separate PR for the benchmark 👍🏻 

then we can create a bunch of scenarios... like what if we have no headers, with and without a matching user-agent value, what about common vs uncommon variants (`HTTP_USER_AGENT`, for example), crazy large set of headers with/without matches, etc, etc...

thanks!

Reply:
Benchmark results of this branch https://github.com/DataDog/dd-trace-py/pull/4047

```
+-------------------------------------------+----------+--------------------------+
| Benchmark                                 | /1.4.0   | /artifacts/2372bcb0-f1b1 |
+===========================================+==========+==========================+
| sethttpmeta-no-useragentvariant           | 251 ms   | 227 ms: 1.11x faster     |
+-------------------------------------------+----------+--------------------------+
| sethttpmeta-all-disabled                  | 86.8 ms  | 89.7 ms: 1.03x slower    |
+-------------------------------------------+----------+--------------------------+
| sethttpmeta-useragentvariant_exists_1     | 228 ms   | 243 ms: 1.06x slower     |
+-------------------------------------------+----------+--------------------------+
| sethttpmeta-useragentvariant_exists_2     | 255 ms   | 261 ms: 1.02x slower     |
+-------------------------------------------+----------+--------------------------+
| sethttpmeta-useragentvariant_exists_3     | 230 ms   | 243 ms: 1.05x slower     |
+-------------------------------------------+----------+--------------------------+
| sethttpmeta-useragentvariant_not_exists_1 | 231 ms   | 235 ms: 1.02x slower     |
+-------------------------------------------+----------+--------------------------+
| sethttpmeta-useragentvariant_not_exists_2 | 232 ms   | 235 ms: 1.02x slower     |
+-------------------------------------------+----------+--------------------------+
| Geometric mean                            | (ref)    | 1.01x slower             |
+-------------------------------------------+----------+--------------------------+
```